map/map_tf_generator/nodes/map_tf_generator.cpp

```diff
@@ -13,65 +13,84 @@
  * limitations under the License.
  */
+#include <pcl/point_cloud.h>
 #include <pcl/point_types.h>
-#include <pcl_ros/point_cloud.h>
-#include <ros/ros.h>
-#include <sensor_msgs/PointCloud2.h>
 #include <tf2/LinearMath/Quaternion.h>
 #include <tf2_ros/static_transform_broadcaster.h>
+#include <rclcpp/rclcpp.hpp>
+#include <sensor_msgs/msg/point_cloud2.hpp>
+
+#include <pcl_conversions/pcl_conversions.h>
 
 typedef pcl::PointCloud<pcl::PointXYZ> PointCloud;
-std::string map_frame_ = "map";
-std::string viewer_frame_ = "viewer";
 
-void Callback(const PointCloud::ConstPtr & clouds)
+class MapTFGenerator : public rclcpp::Node
 {
-  const unsigned int sum = clouds->points.size();
-  double coordinate[3] = {0, 0, 0};
-  for (int i = 0; i < sum; i++) {
-    coordinate[0] += clouds->points[i].x;
-    coordinate[1] += clouds->points[i].y;
-    coordinate[2] += clouds->points[i].z;
+public:
+  MapTFGenerator() : Node("map_tf_generator")
+  {
+    map_frame_ = declare_parameter("map_frame", "map");
+    viewer_frame_ = declare_parameter("viewer_frame", "viewer");
+
+    sub_ = create_subscription<sensor_msgs::msg::PointCloud2>(
+      "pointcloud_map", rclcpp::QoS{1},
+      std::bind(&MapTFGenerator::Callback, this, std::placeholders::_1));
+
+    static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(
+      std::shared_ptr<rclcpp::Node>(this, [](auto) {}));
   }
-  coordinate[0] = coordinate[0] / sum;
-  coordinate[1] = coordinate[1] / sum;
-  coordinate[2] = coordinate[2] / sum;
 
-  geometry_msgs::TransformStamped static_transformStamped;
-  static_transformStamped.header.stamp = ros::Time::now();
-  static_transformStamped.header.frame_id = map_frame_;
-  static_transformStamped.child_frame_id = viewer_frame_;
-  static_transformStamped.transform.translation.x = coordinate[0];
-  static_transformStamped.transform.translation.y = coordinate[1];
-  static_transformStamped.transform.translation.z = coordinate[2];
-  tf2::Quaternion quat;
-  quat.setRPY(0, 0, 0);
-  static_transformStamped.transform.rotation.x = quat.x();
-  static_transformStamped.transform.rotation.y = quat.y();
-  static_transformStamped.transform.rotation.z = quat.z();
-  static_transformStamped.transform.rotation.w = quat.w();
+private:
+  std::string map_frame_;
+  std::string viewer_frame_;
+  rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr sub_;
 
-  static tf2_ros::StaticTransformBroadcaster static_broadcaster;
-  static_broadcaster.sendTransform(static_transformStamped);
+  std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;
 
-  ROS_INFO_STREAM(
-    "broadcast static tf. map_frame:" << map_frame_ << ", viewer_frame:" << viewer_frame_
-                                      << ", x:" << coordinate[0] << ", y:" << coordinate[1]
-                                      << ", z:" << coordinate[2]);
-}
+  void Callback(const sensor_msgs::msg::PointCloud2::ConstSharedPtr clouds_ros)
+  {
+    PointCloud * clouds;
```

Comment:
No memory is allocated to the pointer.
Either use PointCloud object (not pointer) or allocate memory by
```
PointCloud * cloudes = new PointCloud();
```

Reply:
Thank you! replaced them with just an object.